pkg/provision/cleanup.go

```diff
@@ -0,0 +1,234 @@
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+			log.Info().Msgf("skipping subvolume %s", subvol.Path)
+		}
+	}
+
+	return nil
+}
+
+func checkReservationToDelete(path string, cl *client.Client) bool {
+	log.Info().Msgf("checking explorer for reservation: %s", path)
+	reservation, err := cl.Workloads.NodeWorkloadGet(path)
+	if err != nil {
+		var hErr client.HTTPError
+		if ok := errors.As(err, &hErr); ok {
+			resp := hErr.Response()
+			// If reservation is not found it should be deleted
+			if resp.StatusCode == 404 {
+				return true
+			}
+		}
+		return false
+	}
+
+	if reservation.GetNextAction() == workloads.NextActionDelete {
+		log.Info().Msgf("subvolume with path %s has next action to delete", path)
+		return true
+	}
+
+	return false
+}
+
+// checks running containers for subvolumes that might need to be saved because they are used
+// and subvolumes that might need to be deleted because they have no attached container anymore
+func checkContainers(msgBrokerCon string) (map[string]struct{}, map[string]struct{}, error) {
+	toSave := make(map[string]struct{})
+	toDelete := make(map[string]struct{})
+	zbus, err := zbus.NewRedisClient(msgBrokerCon)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client, err := containerd.New(containerdSock)
+	if err != nil {
+		log.Err(err).Msgf("failed to create containerd connection")
+		return nil, nil, err
+	}
+
+	ns, err := client.NamespaceService().List(context.Background())
+	if err != nil {
+		log.Err(err).Msgf("failed to list namespaces")
+		return nil, nil, err
+	}
+
+	for _, ns := range ns {
+		log.Info().Msgf("Checking namespace %s", ns)
+		ctx := namespaces.WithNamespace(context.Background(), ns)
+		crts, err := client.Containers(ctx, "")
+		if err != nil {
+			log.Err(err).Msgf("failed to list containers")
+			return nil, nil, err
+		}
+
+		var zdbNamespaces []string
+
+		for _, ctr := range crts {
+			log.Info().Msgf("container ID %s", ctr.ID())
+			if ns == "zdb" {
+				zdbCl := initZdbConnection(ctr.ID())
+				defer zdbCl.Close()
+				if err := zdbCl.Connect(); err != nil {
+					log.Err(err).Msgf("failed to connect to 0-db: %s", ctr.ID())
+					continue
+				}
+
+				zdbNamespaces, err = zdbCl.Namespaces()
+				if err != nil {
+					log.Err(err).Msg("failed to retrieve zdb namespaces")
+					continue
+				}
+			}
+
+			spec, err := ctr.Spec(ctx)
+			if err != nil {
+				log.Err(err).Msg("failed to container spec")
+				continue
+			}
+
+			toSave[filepath.Base(spec.Root.Path)] = struct{}{}
+			for _, mnt := range spec.Mounts {
+				if !strings.HasPrefix(mnt.Source, "/mnt/") {
+					continue
+				}
+
+				if len(ns) == 1 && zdbNamespaces[0] == "default" {
+					err := common.DeleteZdbContainer(pkg.ContainerID(ctr.ID()), zbus)
+					if err != nil {
+						log.Err(err).Msgf("failed to delete zdb container %s", ctr.ID())
+						continue
+					}
+					toDelete[filepath.Base(mnt.Source)] = struct{}{}
+				} else {
```

Comment:
I don't see where you mark volumes that belongs to containers that are NOT zdb here. shouldn't these go to the `toSave` map ?